google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/ITAbstractSpannerTest.java

```diff
@@ -0,0 +1,390 @@
+    public void setOnlyInjectOnce(boolean value) {
+      this.onlyInjectOnce = value;
+    }
+
+    protected boolean shouldAbort(String statement, ExecutionStep step) {
+      return probability > random.nextDouble();
+    }
+
+    @Override
+    public void intercept(ParsedStatement statement, StatementExecutionStep step,
+        UnitOfWork transaction) {
+      if (shouldAbort(statement.getSqlWithoutComments(), ExecutionStep.of(step))) {
+        // ugly hack warning: inject the aborted state into the transaction manager to simulate an
+        // abort
+        if (transaction instanceof ReadWriteTransaction) {
+          try {
+            Field field = ReadWriteTransaction.class.getDeclaredField("txManager");
+            field.setAccessible(true);
+            TransactionManager tx = (TransactionManager) field.get(transaction);
+            Class<?> cls = Class.forName("com.google.cloud.spanner.TransactionManagerImpl");
+            Class<?> cls2 =
+                Class.forName("com.google.cloud.spanner.SessionPool$AutoClosingTransactionManager");
+            Field delegateField = cls2.getDeclaredField("delegate");
+            delegateField.setAccessible(true);
+            TransactionManager delegate = (TransactionManager) delegateField.get(tx);
+            Field stateField = cls.getDeclaredField("txnState");
+            stateField.setAccessible(true);
+
+            // first rollback, and then pretend it aborted
+            tx.rollback();
+            stateField.set(delegate, TransactionState.ABORTED);
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+          if (onlyInjectOnce) {
+            probability = 0;
+          }
+          throw SpannerExceptionFactory.newSpannerException(ErrorCode.ABORTED,
+              "Transaction was aborted by interceptor");
+        }
+      }
+    }
+  }
+
+  public static String getHost() {
+    return System.getProperty(GceTestEnvConfig.GCE_SERVER_URL, CLOUDSPANNER_HOST);
+  }
+
+  public static String getKeyFile() {
+    return System.getProperty(GceTestEnvConfig.GCE_CREDENTIALS_FILE, DEFAULT_KEY_FILE);
+  }
+
+  public static String getProject() {
+    return System.getProperty(GceTestEnvConfig.GCE_PROJECT_ID, DEFAULT_PROJECT_PLACEHOLDER);
+  }
+
+  protected static Spanner getSpanner() {
+    return spanner;
+  }
+
+  protected static DatabaseId getDatabaseId() {
+    return databaseId;
+  }
+
+  protected DatabaseClient getDatabaseClient() {
+    return spanner.getDatabaseClient(getDatabaseId());
+  }
+
+  protected DatabaseAdminClient getDatabaseAdminClient() {
+    return spanner.getDatabaseAdminClient();
+  }
+
+  @BeforeClass
+  public static void setup() throws IOException, InterruptedException, ExecutionException {
+    createSpanner();
+    createInstance();
+    createDatabase();
+  }
+
+  private static void createSpanner() throws IOException {
+    // generate a unique instance id for this test run
+    Random rnd = new Random();
```

Comment:
can we use the `IntegrationTestEnv` that already exists in the Spanner client lib tests?

Reply:
Yes, I think that is doable. I'll look into it.